packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     ];
   }
 }
+
+class _RenderEditableCustomPaint extends RenderBox {
+  _RenderEditableCustomPaint({
+    RenderEditablePainter? painter,
+  }) : _painter = painter,
+       super();
+
+  @override
+  RenderEditable? get parent => super.parent as RenderEditable?;
+
+  @override
+  bool get isRepaintBoundary => true;
+
+  @override
+  bool get sizedByParent => true;
+
+  RenderEditablePainter? get painter => _painter;
+  RenderEditablePainter? _painter;
+  set painter(RenderEditablePainter? newValue) {
+    if (newValue == painter)
+      return;
+
+    final RenderEditablePainter? oldPainter = painter;
+    _painter = newValue;
+
+    if (newValue?.shouldRepaint(oldPainter) ?? true)
+      markNeedsPaint();
+
+    if (attached) {
+      oldPainter?.removeListener(markNeedsPaint);
+      newValue?.addListener(markNeedsPaint);
+    }
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    final RenderEditable? parent = this.parent;
+    assert(parent != null);
+    final RenderEditablePainter? painter = this.painter;
+    if (painter != null && parent != null) {
+      painter.paint(context.canvas, Size.zero, parent);
+    }
+  }
+
+  @override
+  void attach(PipelineOwner owner) {
+    super.attach(owner);
+    _painter?.addListener(markNeedsPaint);
+  }
+
+  @override
+  void detach() {
+    _painter?.removeListener(markNeedsPaint);
+    super.detach();
+  }
+
+  @override
+  Size computeDryLayout(BoxConstraints constraints) => constraints.biggest;
+}
+
+/// An interface that paints within a [RenderEditable]'s bounds, above or
+/// beneath its text content.
+///
+/// This painter is typically used for painting auxiliary content that depends
+/// on text layout metrics (for instance, for painting carets and text highlight
+/// blocks). It can paint independently from its [RenderEditable], allowing it
+/// to repaint without triggering a repaint on the entire [RenderEditable] stack
+/// when only auxiliary content changes (e.g. a blinking cursor) are present. It
+/// will be scheduled to repaint when:
+///   * It's assigned to a new [RenderEditable] and the [shouldRepaint] method
+///     returns true.
+///   * Any of the [RenderEditable]s it is attched to repaints.
+///   * The [notifyListeners] method is called, which typically happens when the
+///     painter's attributes change.
+///
+/// See also:
+///  * [RenderEditable.setForegroundPainter], which takes a [RenderEditablePainter]
+///    and sets it as the foreground painter of the [RenderEditable].
+///  * [RenderEditable.setPainter], which takes a [RenderEditablePainter]
+///    and sets it as the foreground painter of the [RenderEditable].
+///  * [CustomPainter] a similar class which paints within a [RenderCustomPaint].
+abstract class RenderEditablePainter extends ChangeNotifier {
```

Comment:
The painter doesn't have `hitTest` or `semanticsBuilder` because these currently don't seem to have a use case.

Reply:
The `RenderEditable` property was removed because you can use the same painter instance in different render objects simultaneously when the tree finalizes (for example in a `Hero` widget we'd keep 2 copies of the subtree so in case there's a painter in the subtree it will be attached to 2 different `RenderEditable`s).